### PR TITLE
hv: Add x86 TEE support

### PR DESCRIPTION
**Pull request by yfliuuu**
This PR contains patches that fixes the v4 of Deng Jie's x86_tee patch series.

One minor change from his original version: 

In hv: tee: Support the concept of companion VM, a default value of 65535 is added as companion_vmid for successful compilation.

### CODE REVIEW

File: hypervisor/include/arch/x86/asm/guest/optee.h [hunk 1]
```diff
@@ -9,10 +9,17 @@
...
 #include <asm/vm_config.h>
+#include <ptdev.h>
+
+#define TEE_NOTIFICATION_VECTOR 0x29U
 
 /* If the RDI equals to this value, then this is a RETURN after FIQ DONE */
 #define OPTEE_RETURN_FIQ_DONE	0xBE000006UL
 
+/* This value tells OPTEE that this switch to TEE is due to secure interrupt */
+#define OPTEE_SECURE_INT_INDICATOR	0xB20000FFUL
```
Comment (ywan170) [inline hypervisor/include/arch/x86/asm/guest/optee.h:20]: Suggest to match the OPTEE_RETURN_FIQ_DONE. Rename to OPTEE_FIQ_ENTRY.

File: hypervisor/arch/x86/guest/optee.c [hunk 2]
```diff
@@ -145,3 +145,40 @@ int32_t hcall_switch_ee(struct acrn_vcpu *vcpu, __unused struct acrn_vm *target_
...
+	 * injected to TEE directly. But when REE vcpu is running
+	 * at that time, we need to switch to TEE for handling,
+	 * and copy 0xB20000FF to RDI to notify OPTEE about this.
+	 */
+	if (((get_vm_config(entry->vm->vm_id)->guest_flags & GUEST_FLAG_REE) != 0U) &&
+		((get_vm_config(curr_vcpu->vm->vm_id)->guest_flags & GUEST_FLAG_TEE) != 0U)) {
+		tee_vcpu = vcpu_from_pid(get_companion_vm(entry->vm), pcpu_id);
+		vlapic_set_intr(tee_vcpu, TEE_NOTIFICATION_VECTOR, LAPIC_TRIG_EDGE);
+	} else if (((get_vm_config(entry->vm->vm_id)->guest_flags & GUEST_FLAG_TEE) != 0U) &&
+		((get_vm_config(curr_vcpu->vm->vm_id)->guest_flags & GUEST_FLAG_REE) != 0U)) {
```
Comment (dongyaozu) [inline hypervisor/arch/x86/guest/optee.c:172]: Ditto

File: hypervisor/arch/x86/guest/optee.c [hunk 3]
```diff
@@ -145,3 +145,40 @@ int32_t hcall_switch_ee(struct acrn_vcpu *vcpu, __unused struct acrn_vm *target_
...
+	 * to REE directly. But when TEE vcpu is running at that time,
+	 * we need to inject a predefined vector to TEE for notification
+	 * and continue to switch back to TEE for running.
+	 *
+	 * Secure interrupt (interrupt belongs to TEE) comes
+	 * when TEE vcpu is running, the interrupt will be
+	 * injected to TEE directly. But when REE vcpu is running
+	 * at that time, we need to switch to TEE for handling,
+	 * and copy 0xB20000FF to RDI to notify OPTEE about this.
+	 */
```
Comment (dongyaozu) [inline hypervisor/arch/x86/guest/optee.c:166]: ......   See next comment.

File: hypervisor/include/arch/x86/asm/guest/optee.h [hunk 4]
```diff
@@ -10,6 +10,9 @@
 #include <asm/guest/vm.h>
 #include <asm/vm_config.h>
 
+/* If the RDI equals to this value, then this is a RETURN after FIQ DONE */
+#define OPTEE_RETURN_FIQ_DONE	0xBE000006UL
```
Comment (dongyaozu) [inline hypervisor/include/arch/x86/asm/guest/optee.h:17]: Where does the magic # come from? If it is from a spec, please identify the source here.

Comment (ywan170) [reply hypervisor/include/arch/x86/asm/guest/optee.h:17]: This is from OPTEE. Standard OPTEE command.

**Comment by yfliuuu** [comment]
Re-pushed: 
Rename macro OPTEE_SECURE_INT_INDICATOR to OPTEE_FIQ_ENTRY
Break the comments in optee.c and place them in respective branches
Add @ywan170 's Reviewed-by

**Review by dongyaozu — APPROVED** [review]

**Comment by wenlingz** [comment]
No new violations to the coding guideline detected.
No New Name Conflict
No New Function Declaration/Definition Mismatch

**Pull request merged**